include/boost/geometry/util/promote_integral.hpp

```diff
@@ -0,0 +1,276 @@
+>
+class promote_integral
+{
+private:
+    static bool const is_unsigned = boost::is_unsigned<T>::type::value;
+
+#if !defined(BOOST_GEOMETRY_NO_MULTIPRECISION_INTEGER)
+    // Define the proper check policy for the multiprecision integer
+    typedef typename boost::mpl::if_c
+        <
+            UseCheckedMultiprecisionInteger,
+            boost::integral_constant
+                <
+                    boost::multiprecision::cpp_int_check_type,
+                    boost::multiprecision::checked
+                >,
+            boost::integral_constant
+                <
+                    boost::multiprecision::cpp_int_check_type,
+                    boost::multiprecision::unchecked
+                >
+        >::type check_policy_type;
+
+    // Meta-function to get the multiprecision integer type for the
+    // given size and type (signed/unsigned)
+    template
+    <
+        unsigned int Size,
+        boost::multiprecision::cpp_integer_type IntegerType
+    >
+    struct multiprecision_integer_type
+    {
+        typedef boost::multiprecision::number
+            <
+                boost::multiprecision::cpp_int_backend
+                    <
+                        Size,
+                        Size,
+                        IntegerType,
+                        check_policy_type::value,
+                        void
+                    >
+            > type;
+    };
+
+    // Meta-function to get the multiprecision signed integer type for
+    // the given size
+    template <unsigned int Size>
+    struct multiprecision_signed_integer_type
+    {
+        typedef typename multiprecision_integer_type
+            <
+                Size, boost::multiprecision::signed_magnitude
+            >::type type;
+    };
+
+    // Meta-function to get the multiprecision unsigned integer type for
+    // the given size
+    template <unsigned int Size>
+    struct multiprecision_unsigned_integer_type
+    {
+        typedef typename multiprecision_integer_type
+            <
+                Size, boost::multiprecision::unsigned_magnitude
+            >::type type;
+    };
+
+    // Define the minimum (and maximum) bit size for the
+    // multiprecision integer needed
+    // If T is the input type and P the promoted type, then the
+    // minimum number of bits for P are (below b stands for the number
+    // of bits of T):
+    // * if T is unsigned and P is unsigned: 2 * b
+    // * if T is signed and P is signed: 2 * b - 1
+    // * if T is unsigned and P is signed: 2 * b + 1
+    typedef typename boost::mpl::if_c
+        <
+            (PromoteUnsignedToUnsigned && is_unsigned),
+            boost::integral_constant
+                <
+                    unsigned int, (2 * CHAR_BIT * sizeof(T))
+                >,
+            boost::integral_constant
+                <
+                    unsigned int,
+                    (2 * CHAR_BIT * sizeof(T) + (is_unsigned ? 1 : -1))
```

Comment:
Shouldn't here `std::size_t` be used as below?

`boost::mpl::size_t<>` could be used instead of integral constant. It'd be more clear.

For signed numbers number of bits is decreased by 1 (`-1`). Shouldn't it be the same?

I'm wondering, what's the representation of a number in MultiPrecision? I'm assuming that the space isn't increased by one bit but something bigger, e.g. a byte or more and the number is truncated somehow. I'm guessing that this can add an overhead. Wouldn't it be better to add the whole byte for unsigned or even multiply the `size` by 3? So e.g.: `CHAR_BIT * (2 * sizeof(T) + (is_unsigned ? 1 : 0))`. Or maybe even better find the next power of 2 dividable by CHAR_BIT (which can be greater than 8). In case if some optimization was done in MultiPrecision, i.e. numbers stored on an array of types as big as possible or something like that. So e.g. for 128b unsigned, use 256b promoted signed. Depending on the representation. Anyway, it could be added later.